### DISCUSSION TITLE
Debugging Help

### DISCUSSION THREAD

**Discussion by lexiekark** [perspective-dev/perspective#1967] [Q&A] [answered]
Hello,

I recently started editing the perspective-viewer rust code (in the components directory) to fix the "in" and "not in" filters.  

I was wondering how people go about setting breakpoints/debugging the perspective-viewer component code? I think I was able to make the filters work for the most part via trial and error.  Now I want to see what portions of code are running when I focus or click out of the filters.  I tried to put some print statements in the functions, to see what order everything was being called in and with what inputs, but those didn't show up in the console. Do people usually add breakpoints within Google Chrome? I looked in my sources tab and I wasn't sure if that was the right way to go either.

Thank you in advance for any suggestions on how to contribute to this code more effectively!

**Comment by texodus** [comment, marked as answer]
A couple of tips -
* There is no sourcemap support for Rust's `wasm32-unknown-unknown` that I am aware of, so debugging is limited to breakpoints in the wasm bytecode or JS, neither of which are particularly useful for UX bugs.  The latter will emit readable sourcemaps in debug builds at least.
* `println!()` does not work in WASM as there is no `stdout`.  You can log to the JS console with the [`tracing` crate](https://docs.rs/tracing/latest/tracing/), but you'll need to make sure you're building `perspective-viewer` in debug mode (configured via the `yarn setup` menu), otherwise [logs below `error` will be omitted from the build](https://github.com/finos/perspective/blob/master/rust/perspective-viewer/Cargo.toml#L76).

**Reply by lexiekark** [reply]
Thanks so much, will give this a try!
